File: app/src/lib/widgetSync.ts
import { Platform } from "react-native";
import { pb, PB_URL } from "./pb";
import { PearShared } from "../../modules/pear-shared/src";

/**
 * Issues a revocable widget token from the server and hands it (plus the API
 * base URL) to the WidgetKit extension via the shared App Group container,
 * then asks WidgetKit to reload timelines.
 *
 * Best-effort by design: the app works fine without the widget installed.
 */
export async function syncWidgetCredentials(): Promise<void> {
  if (Platform.OS !== "ios") return;
  try {
    const res = await fetch(`${PB_URL}/api/peard/widget/token`, {
      method: "POST",
      headers: { Authorization: pb.authStore.token },
    });
    if (!res.ok) return;
    const data = (await res.json()) as { token: string };
    PearShared.setSharedItem("widgetToken", data.token);
    PearShared.setSharedItem("apiBaseUrl", PB_URL);
    PearShared.reloadTimelines();
  } catch {
    // no-op: widget sync is opportunistic
  }
}

export function clearWidgetCredentials(): void {
  if (Platform.OS !== "ios") return;
  try {
    PearShared.removeSharedItem("widgetToken");
    PearShared.reloadTimelines();
  } catch {
    // no-op
  }
}

File: ios/Peard/WidgetSync.swift
import Foundation
import PeardCore
import WidgetKit

/// Issues the widget token and hands it to the App Group container
/// (Requirement 16).
@MainActor
final class WidgetSync {
    private let api: APIClient
    private let store: SharedStore
    private let baseURL: URL

    init(api: APIClient, store: SharedStore, baseURL: URL) {
        self.api = api
        self.store = store
        self.baseURL = baseURL
    }

    /// Best-effort by design: a failure leaves the container untouched and the
    /// session alive (Requirement 16.3).
    func sync() async {
        do {
            let issue = try await api.issueWidgetToken()
            store.writeWidgetCredentials(token: issue.token, baseURL: baseURL)
            reloadTimelines()
        } catch {
            // Widget sync is opportunistic; the app works without the widget.
        }
    }

    /// Removes the token and refreshes timelines (Requirement 16.4).
    func clear() {
        store.removeWidgetToken()
        reloadTimelines()
    }

    func reloadTimelines() {
        WidgetCenter.shared.reloadAllTimelines()
    }
}
